framework/include/utils/LibtorchArtificialNeuralNetTrainer.h

```diff
@@ -0,0 +1,308 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#ifdef LIBTORCH_ENABLED
+
+#include "LibtorchNeuralNet.h"
+#include "LibtorchDataset.h"
+#include "DataIO.h"
+#include "MooseEnum.h"
+
+namespace Moose
+{
+
+/**
+ * A struct containing necessary information for training neural networks.
+ * This is an easy way to pass multiple parameters to the neural networks
+ * and follows the design of multiple libtorch classes.
+ */
+struct LibtorchTrainingOptions
+{
+  /// The type of optimizer we want to use for training, adam is the default due to
+  /// its robustness and fast convergence
+  MooseEnum optimizer_type = MooseEnum("adam adagrad rmsprop sgd", "adam");
+  /// Number of iterations we want to perform on the whole dataset
+  unsigned int num_epochs = 1;
+  /// Number of batches we want to split the dataset into
+  unsigned int num_batches = 1;
+  /// If we want to print additional information during training
+  bool print_loss = false;
+  /// The frequency of training loss print to console
+  unsigned int print_epoch_loss = 1;
+  /// The relative loss tolerance where the training shall stop
+  Real rel_loss_tol = 1e-12;
+  /// The learning rate for the optimizers
+  Real learning_rate = 1e-3;
+  /// The number of allowed parallel processes. It is necessary to bound it because the user may
+  /// want to run simulations with more processors than data samples which leads to segmentation faults
+  /// is a process remains with no sample
+  unsigned int parallel_processes = 1;
+  /// Parameter for sampling. This actually allows random samplers to use the same sample
+  /// multiple times. Sometimes comes in handy when only a few of them are needed to fill up
+  /// empty slots in the last batch. NOTE: This may introduce a considerable BIAS!
+  bool allow_duplicates = false;
+};
+
+/**
+ * Templated class which is responsible for training LibtorchArtificialNeuralNets. The specialty
+ * of this class is that it can be used with random/sequential samplers in serial or
+ * parallel. The default sampling approach is serial.
+ */
+template <typename Sampler = torch::data::samplers::DistributedSequentialSampler>
+class LibtorchArtificialNeuralNetTrainer
```

Comment:
Use explicit instantiation to get this out of a header file

Reply:
I explicitly instantiate this for `torch::data::samplers::DistributedSequentialSampler` and `torch::data::samplers::DistributedRandomSampler` in `LibtorchArtificialNeuralNetTrainer.C` now.